File: 15-3sum/3sum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
    let result = [];
    nums.sort((a, b) => a - b);
    for(let i = 0; i < nums.length; i++){
        if (i > 0 && nums[i] == nums[i - 1]) continue; 
        let j = i + 1;
        let k = nums.length - 1 ;
        
        while(j < k){
            let sum = nums[i] + nums[j] + nums[k]
            if(sum > 0){
                k--;
            }
            else if( sum < 0 ){
                j++;
            }
            else{
                result.push([nums[i], nums[j], nums[k]]);
                while(j<k && nums[j] === nums[j+1]) j++;
                j++;            }
        }
    }
    return result;
};

File: 15-3sum/3sum.py
class Solution:
    def threeSum(self, nums: list[int]) -> list[list[int]]:
        nums.sort()
        result = []
        n = len(nums)
        for i in range(n - 2):
            if i > 0 and nums[i] == nums[i - 1]:
                i +=1
                continue
            
            j = i + 1
            k = n - 1

            while j < k:
                sum = nums[i] + nums[j] + nums[k]
                if sum == 0:
                    result.append([nums[i], nums[j], nums[k]])
                    j += 1
                    k -= 1

                    # Skip duplicate second elements
                    while j < k and nums[j] == nums[j - 1]:
                        j += 1

                    # Skip duplicate third elements
                    while j < k and nums[k] == nums[k + 1]:
                        k -= 1
                elif sum > 0:
                    k -= 1
                else:
                    j += 1

        return result
                
